Finish the experimental math expression to blocks translator

BookOwl opened the pull request:
As @brianharvey asked, I made the experimental math parser follow PEMDAS and be left-associative.

Screenshots:
<img width="188" alt="1" src="https://cloud.githubusercontent.com/assets/13385064/21356707/9b8dad02-c6a0-11e6-9ffe-bf16621818b2.png">
<img width="235" alt="untitled" src="https://cloud.githubusercontent.com/assets/13385064/21361195/eae30dba-c6b0-11e6-975b-d84e35bfa5d2.png">
<img width="190" alt="3" src="https://cloud.githubusercontent.com/assets/13385064/21356708/9b925b04-c6a0-11e6-97e8-d0fd6f4ea99d.png">

BookOwl committed:
Work on the new parser

BookOwl committed:
Finished the experimental parser.
Fixed a bug in the AST to block generator.

BookOwl committed:
Remove debug console.log calls

tjvr commented:
This is in conflict :( Can you try rebasing onto the latest master?

tjvr commented on objects.js:
```diff
@@ -2640,87 +2640,135 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
+                tokens.push({ type: "lp", val: "("});
+                expressionString = expressionString.slice(1);
+            } else if (expressionString[0] === ")") {
+                tokens.push({ type: "rp", val: "("});
```
Why `val (` here?

BookOwl replied:
Oops, typo. It doesn't really matter, but I'll fix it.

tjvr commented on objects.js:
```diff
@@ -2640,87 +2640,135 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
+                            throw "No comma between arguments";
+                        }
+                    }
+                    tokens.splice(0, 1);
```
Why not `tokens.shift()` instead? :)

BookOwl replied:
Because MDN said that you use splice to remove elements. My JS is kind of rusty.

BookOwl replied:
I know, but it also returns it. Since I'm not doing anything with the result, I thought that splice was better. I'll change it if you think that shift is better though.

BookOwl replied:
I changed it just for you, @tjvr

tjvr commented:
This looks plausible to me, FWIW :)

BookOwl committed:
Fixed a teeny tiny typo

BookOwl committed:
Merge upstream commit

BookOwl commented:
Yeah, conflicts are gone!

BookOwl committed:
Changed token.splice(0, 1) into token.shift()
(now blob can be happy)

brianharvey commented:
Thanks, Matthew, for the programming, and Tim, for the code review!

I don't quite understand the second of three screenshots in the first post on this page, though.  How did 5*7/x turn into 5*7*x?  Or was that the teeny typo?

BookOwl commented:
Nope, that's a different bug. Thanks for pointing it out.

BookOwl committed:
Fixed the parsing of sequences of operations that have the same
precedence, like "5 * 4 / 6" and "89 - 12 + 6 - 7"

Thank you @brianharvey for finding this bug

BookOwl commented:
Fixed

cycomachead commented on objects.js:
```diff
@@ -2640,87 +2640,137 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
+        }
+        function findNumber() {
+            var next = "";
+            while (expressionString.length > 0 && /\d|\./.test(expressionString[0])) {
```
Our strings are probably not long enough to make a huge difference, but `.slice()` allocates a new string, so it seems better to just use an index.

Also why check `.length > 0` when `expressionString` is false if empty.

BookOwl replied:
I am using an index...
I'll remove the .length check

cycomachead replied:
(Apparently, I never hit post on this 'review'...you've already addressed some of the comments so sorry for the duplication. 

Sorry for not being clear, what I meant was an explicit index variable. (Maybe `position` or something, that is just an int.)
Then each loop would simply do `position += 1` and checking `expressionString[0]` would be `expressionString[position]`.

I don't know if it really makes a difference though, but I imagine with a long string and parsing on every change it could be noticeable especially on slower computers.

cycomachead replied:
It might also make sense to set `expressionString[0]` to an explicit variable, but I don't know if there's much difference. (in readability or perf.)

cycomachead commented on objects.js:
```diff
@@ -2640,87 +2640,137 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
+        }
+        while (expressionString.length > 0) {
+            expressionString = expressionString.trim();
+            if (/[A-Za-z_]/.test(expressionString[0])) {
```
I feel like all regex's should be defined and given a semantic name, otherwise it's difficult to read why they are used.

tjvr replied:
Personally I find that harder to follow — then I have to look elsewhere to work out what it's doing! But that's just me :-)

BookOwl replied:
I'll change this since Jens agrees with @cycomachead

cycomachead replied:
What it _is_ doing and what it is _trying_ do are two separate things. The reason to name things is because they help convey programmer intent, not just function. :) When I look at this I question whether or not `-` can be in a variable name or whether this is a complete set of characters. I completely forgot to ask that question before....

cycomachead commented on objects.js:
```diff
@@ -2640,87 +2640,137 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
-            return [operator, already, format(inputs[1])];
+        function precedence1(tokens) {
+            var exp = precedence0(tokens);
+            if (tokens.length === 0) {
```
Typically this would just by `!tokens`

tjvr replied:
Nope, only if tokens were a list. The empty list is not falsey in JS! This isn't Python... :-)

liam4 replied:
Typically this would just be !tokens.length from what I've seen in other code bases. But I prefer `.length === 0`.

cycomachead replied:
OMG Github just lost my comment. :(

Anyway... You're right! I've been not writing enough JS lately....

I was suggesting `!tokens.length` then, because that's more common in Snap!'s code too. But I don't have a strong preference...

(In Snap! .length is used as a boolean about 60-70 times, and `=== 0` only 15, according to my basic grepping.)

jmoenig commented:
Thanks! Let me know when you feel it's ready for heavy testing.

Can you make it support negative numbers? Currently your parser turns negative numbers into separate nodes subtracting the number from zero. While this is kinda correct mathematically it makes the resulting formula much more complex than it needs to be.
Thanks!

cycomachead commented on objects.js:
```diff
@@ -2640,87 +2640,135 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
     var ast;
 
-    function parseInfix(expression, operator, already) {
-        // very basic binary infix parser for arithmetic expressions
```
This comment should be retained, though in an updated form, I think - including what happens when an error is thrown.

BookOwl replied:
Which function should I put the comment in?

cycomachead replied:
I don't know if it matters much, but probably around createAST since that's doing essentially the same work.

cycomachead commented on objects.js:
```diff
@@ -2640,87 +2640,135 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
+        var tokens = [];
+        var token;
+        expressionString = expressionString.trim();
+        function findIdent() {
```
Maybe I'm daft, but I am confused by the name `Indent`. What is meant by Indent? It seems to be anything that's text?

tjvr replied:
Ident is short for identifier. You've added an N :-)

BookOwl replied:
Oops, another typo. I'll fix it.

cycomachead replied:
That may have been autocorrect. :) In any case, then it should be `findIdentifer` We're not saving that many characters, and my first guess was actually "Identity" but that didn't make sense.

cycomachead commented on objects.js:
```diff
@@ -2779,7 +2829,7 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
 
     if (expressionString > 100) {return null; }
```
More for @jmoenig, but I also feel like these magic numbers should be given meaningful names. :)

Also I'm pretty sure this should be `expressionString.length`?

jmoenig replied:
Oh sure, right!

jmoenig replied:
I just fixed that in the repo

jmoenig replied:
Thanks, Michael!

jmoenig commented:
Hey, @BookOwl , thanks for pointing out the glitch in my code in the ```round``` case! I just pushed a fix which now causes conflicts :-(

How about if you simply exchange my whole ```reporterize``` function with your own one, that way we can keep both in the source for a while without running into conflicts?

jmoenig commented:
Another question to consider is whether we should add a ```raisedTo``` operator/reporter primitive to Snap...

BookOwl commented:
@jmoenig, I can't make `-x` not be `0 - x` in the general case because Snap doesn't have a negate block, so `-(4 * 7)` would have to be turned into `(0 - (4 * 7))`, and I just had numbers do the same thing because I didn't have to do any extra work.

jmoenig commented:
right, but you can make negative numbers, such as -42, and just use them as inputs. No need to turn -42 into (0 - 42). Make sense?

BookOwl commented:
Yes

BookOwl commented:
Should I add the boolean operators in my version of the parser/block generator?

BookOwl committed:
Cleaned up the code by giving the RegExps names and giving some
variables more descriptive names.
Negative numbers are now turned into negative numbers in the generated
blocks, instead of being "0 - num"

BookOwl commented:
I made the negative numbers parse correctly.

BookOwl committed:
Resolve conflicts

BookOwl commented:
I resolved the conflicts be just telling git that my version is what should be kept.

dekrain commented on objects.js:
```diff
@@ -2856,7 +2838,7 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
     }
 
-    if (expressionString.length > 100) {return null; }
+    if (expressionString > 100) {return null; }
```
You removed fix of bug.

BookOwl replied:
Oops, thanks for pointing that out

brianharvey commented:
@jmoenig #raisedto: All these years we've been resisting adding as primitives things that can usefully (as a learning experience) be written in Snap<i>!</i> itself.  Perhaps we should revisit that policy, but not for this reason -- it'd be the tail wagging the dog.

In this case we probably want a full math library with ^ as just one piece, optionally loadable.

brianharvey commented:
@BookOwl #Booleans: Bearing in mind that the original motivation for this whole feature was pasting in formulas from published work, if it's a math paper rather than a computer program, Boolean symbols mean something more like assert(...) than like a computation.  I suppose it'd be possible for ∧ and ∨ to have one meaning while & and | have a different meaning.  But I still think we should live with arithmetic for a while before we start planning extensions.

BookOwl commented:
@brianharvey, I'm asking because @jmoenig has already added it to his version of the parser.

cycomachead reviewed (commented):
I actually do think this is quite clean, but stuffing a whole infix parser inside objects.js feels wrong to me me. I'm not trying to like stake a claim on this, but I do feel like we've passed a point of what is comprehensible in some of our files. :-)

cycomachead commented:
While we are on the topic of design things:
I find `5 * ` to generate `5 x 0` confusing. I will like the 0 should be unfilled until I type something there.

BookOwl commented:
`5 *` doesn't generate `5 x 0`, or anything else. If you want an empty input slot, you write `5 * _`

brianharvey commented:
Oh, well, if Jens has it, by all means.  Might as well have the relational operators too.  Of course this adds more levels of binding strength. :P

cycomachead commented:
> 5 * doesn't generate 5 x 0, or anything else. If you want an empty input slot, you write 5 * _

Oh, interesting. On Jens' version it does:
<img width="289" alt="screen shot 2016-12-21 at 10 48 53 pm" src="https://cloud.githubusercontent.com/assets/1505907/21418487/765747a0-c7d6-11e6-95b0-0b6c5f57aaab.png">

I think this is kinda what I would expect, modulo the 0 being there.

BookOwl committed:
Fix merge conflicts

BookOwl committed:
Cleaned up the code and removed a dubug console.log call

BookOwl committed:
Fix typo in comment

BookOwl commented:
@jmoenig, do you think that the boolean and relational operators should be in here?

jmoenig commented:
@BookOwl thanks for the commits! What's the consensus here? Is this now ready for prime-time, or would you like to keep testing and refining it? (Hey, no rush whatsoever, this is coming along just fine). Which operators would you like to change or add?

BookOwl committed:
Added pick random to the expression parser as random(lower, upper)

BookOwl commented:
I just added pick random, so unless you want me to add the boolean and relational operators, I think that this is ready for you to test.

BookOwl committed:
Add back jens bugfix

cycomachead commented on objects.js:
```diff
@@ -2634,96 +2634,149 @@ SpriteMorph.prototype.searchBlocks = function (
-                expr += ch;
+    function tokenize(expressionString) {
+        // very basic tokenizer for arithmetic expressions
+        var tokens = [];
```
These should be:
```js

var a = 1,
    b = 2,
    c = 3;
```
formatted (to be consistent with Snap<em>!</em>)

Also, I really feel like `IDENT` should just be `IDENTIFIER` since we're not really short on characters.

BookOwl replied:
OK.

cycomachead commented on objects.js:
```diff
@@ -2634,96 +2634,149 @@ SpriteMorph.prototype.searchBlocks = function (
+        // very basic tokenizer for arithmetic expressions
+        var tokens = [];
+        var DIGITS = /\d|\./;
+        var IDENT_START = /[A-Za-z_]/;
```
Is this being used to match variable names? If so it seems like we should greatly expand what it accepts, though that does significant complicate things if it were to match spaces and operators, but perhaps we can ignore that for now...

BookOwl replied:
Yes it does, but that regex is just for the start of variable names. The rest of a variable can also contain digits.

cycomachead commented:
When I start with a negative number, there seems to be an infinite loop. 

I was trying stuff with the expression `5 * -10 + 3` if I did certain things to `-10` like `-10.3.3.3` or `-10.1xxx` sometimes the block generated would be (5 x -10) + 3 and sometimes the +3 would be left off. At some point I deleted something and then got into another infinite loop.

cycomachead commented:
Hmm, consider this example. I'm not sure what's right but I'm highlighting it for others to think about:
"-11.0.0.0"
Jens' version turns this in to `() - (11.0.0.0)` where that's a variable/orange blob.
This version turns it into `-11` into whatever block you'd be in, the rest of the "0.0.0" is just dropped.

Jens' version turns "5 -11-11" into 5 - 11 -11, which this version is 5 - 11. 

Ok, some of these are small issues, but it doesn't seem too hard for students to run into weird cases saying "what the heck is going on?" (in both versions). I'd be really useful to have some documentation about a couple of the edge cases with some expectations for what they do.

BookOwl commented:
The reason that "-11.0.0.0.0" gets turned into "-11" is because that is how [JS's `parseFloat` function works](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseFloat)

BookOwl commented:
`5*-10.3-3` having the `-3` get left off _is_ a bug.

What input is causing infinite loops?

BookOwl committed:
Clean up some code

BookOwl committed:
Resolve merge conflicts

jmoenig commented:
this is coming along nicely, thanks! I keep stumbling across glitches and improvements in my much simpler version, haha! This is fun...

brianharvey commented:
The Official Right Thing about minus signs is that `<expr>-<expr>` and `<expr> - <expr>` are dyadic subtraction, but `<expr> -<expr>` is monadic negation (and therefore has a higher binding strength than anything other than parentheses).  Of course `<operator> - <expr>` is monadic regardless of spacing.  And monadic negation of a literal nonnegative number should just be rendered as a minus sign as part of the number itself.

My vote is that a malformed numeric literal like 3.0.0 should be flagged as an error (in Snap<i>!</i>, not just in the text-to-Snap<i>!</i> translation) but that a literal that starts with a digit but includes letters, such as `2nd player`, should be a legal variable name.  But I don't think there's an Official Right Thing about that.

cycomachead commented:
> The reason that "-11.0.0.0.0" gets turned into "-11" is because that is how JS's parseFloat function works

IMO this is a bug. That means (if I understand correctly) that anything that looks like it starts with a number, but could be a variable gets treated as a number. But it is at least currently consistent with how Snap! handles numbers that are malformed. To me though someone trying to type '11-11-11' is probably indicating a bunch of subtractions.

As far as infinite loops: I don't remember the specific example other than just starting off with `-`. But it didn't take too much playing around to get it to happen. (It might not have been an infinite loop, but basically Safari just hung and I couldn't inspect things so that's my best guess.)

cycomachead reviewed (commented):
Sorry this was supposed to be published a few days ago... This is the cost of having too many tabs.

jmoenig commented:
So, again asking the general consensus about whether this pull request is now ready to be integrated? @BookOwl , do you consider your work to be ready for prime time?
Thanks!

BookOwl commented:
@jmoenig, I want to fix it so that `1-1` is correctly parsed into `(1) - (1)` first.

jmoenig commented:
@BookOwl okay, cool!

BookOwl committed:
Made "1-1" parse correctly
Fixed a bug with block generation for round

BookOwl commented:
@jmoenig, I think that this is ready

jmoenig commented:
Thanks!

jmoenig commented on objects.js:
```diff
@@ -2634,100 +2634,164 @@ SpriteMorph.prototype.searchBlocks = function (
-        return format(inputs[0]);
     }
-
     function blockFromAST(ast) {
```
Can you explain why your ```blockFromAST()``` function is different from my version? Is it because I've been changing it after you forked it?

BookOwl replied:
The only difference is that I haven't added the boolean operators.

jmoenig replied:
Ah, I see. So we can just use my ```blockFromAST()``` function on your ASTs, and you could add the Boolean operators to your parser, right? Or, if that's too much trouble we first parse it your way, and - if that doesn't result in anything useful - mine :)

BookOwl replied:
It should be pretty easy to add the boolean and relational operators, I just need you to tell me what precedence they should have.

jmoenig replied:
Good idea, and good question! Least precedence, I think. But I'm sure @brianharvey knows all about it... :)

brianharvey replied:
https://en.wikipedia.org/wiki/Operators_in_C_and_C%2B%2B#Operator_precedence

BookOwl replied:
Is it OK to use & and | for "and" and "or"?

brianharvey replied:
Oh, there's a question.  People who program in infix expect && and || for the Boolean operators, so you should certainly accept them (so that copy-and-paste works).  It might be okay to accept & and | _also._  On the other hand, we really should add bitand and bitor to the primitive dyadic operators someday.

BookOwl replied:
I was just asking because single character operators are easy.

brianharvey replied:
I understand.  I guess I'm saying "no," at least not _instead of_ && and ||.

BookOwl replied:
That's fine. I've already got && and || working, I just need to get ! (for not) to work

BookOwl replied:
It's working! 😄

cycomachead commented on objects.js:
```diff
@@ -2783,10 +2801,11 @@ SpriteMorph.prototype.reporterize = function (expressionString) {
             '%': 'reportModulus',
-            'random': 'reportRandom'
+            'random': 'reportRandom',
+            'round': 'reportRound'
```
This confuses me. It might well be the right place, but `round` is monadic.

BookOwl replied:
I moved round there because even though it is monadic, it is a different type of block from the other monads, and it is easier to treat it as a diadic function. I should change the comments to make this clear

BookOwl replied:
Done

BookOwl committed:
Add the boolean and relational operators
I also cleaned up the blocksFromAST function

BookOwl commented:
OK, so the boolean and relational operators have been added! 😄 Here are the supported operators and their precedence, from lowest to highest:
```
&&, ||
>, =, <
+, -
*, /
! (a unary operator)
```

BookOwl committed:
Fix for tokenize to keep it from going into an infinite loop on bad expressions

dekrain commented:
Soon 100 comments!

jmoenig commented:
Thanks, @BookOwl , this is looking great now! I'll pull it for the next - upcoming - release, probably sometime next week (the pull, not the release ^^)

BookOwl commented:
Thanks! :D

jmoenig commented:
Okay, looking at this again I'm noticing a few issues. I'll leave this open and figure out how to integrate it, but perhaps not in the upcoming release due in the next days. Hang on!

BookOwl commented:
Is there anything that I can fix?

jmoenig commented:
It's gotten more complicated and a lot bigger than what I was hoping for. I was hoping for something more minimalistic, because I'm basically reluctant to add so much code for a full-fledged parser to ```objects.js```. Perhaps I'll decide to add a new file for parsing stuff to the Snap code base, which is where this could end up in.

BookOwl commented:
OK

brianharvey commented:
In response to

> In any case, I don't know what we can fix that one b/c that's how JS displays numbers. Scientific notation is was what users will see if they were to get that as the result of a math block in Snap!, so it's not totally unexpected.

(posted in #1591 by mistake):

Bignums FTW!!!!

BookOwl commented:
How would bignums work with the rest of Snap!?

brianharvey commented:
We have a bignum library, which Jens won't let me install as a library because he says 4.1 will break it.  Not just bignums, but the complete Scheme numeric tower.  It replaces the arithmetic primitives and then everything just works.
![bignum](https://cloud.githubusercontent.com/assets/3885074/21630647/1a71bac6-d1e8-11e6-832b-5186790814c5.png)
![rat](https://cloud.githubusercontent.com/assets/3885074/21630725/c15b7368-d1e8-11e6-9f60-1d07ecfab7cd.png)
![imag](https://cloud.githubusercontent.com/assets/3885074/21630730/d03c1180-d1e8-11e6-9091-91961976fdc3.png)

BookOwl commented:
I don't think that this should rely on having an optional library, especially if Jens doesn't want the library added.

brianharvey commented:
Oh, sorry, I didn't mean to suggest that you should write code around bignums, it's just that someone said, on that other thread, that nothing can be done about switching representations because that's what JS does, and I'm providing a counterexample.

BookOwl commented:
@jmoenig, would this be more manageable if I broke out the expression parsing and translating code into a new file named `math_translator.js` or similar, and have that file expose a single function that took a string containing a math expression and returned a block, and changed `reporterized` to call that function?

jmoenig commented:
No, I'm not a fan of IIFEs. Let's keep this around until I'm less under fire to include features and change things for the urgently needed release (the release is urgend, because Chrome 55 breaks Snap's music feature, which the release will fix). Eventually I'll get around to find a good way to integrate this.

micahcowan commented:
Right now, function-call identifiers get looked up in a list of `reportMonadic` variants, and then in a mapping of operator names, which get passed to `SpriteMorph.prototype.blockForSelector()`.

Are there any thoughts on, if those two checks fail, trying the key/identifier directly in blockForSelector? i.e., in this block of code from `blockFromAST()`:

```javascript
        } else { // The function or operator has its own block
            block = SpriteMorph.prototype.blockForSelector(selectors[key]);
            inps = block.inputs();
        }
```

the `block =` line becomes something like:

```javascript
            block = SpriteMorph.prototype.blockForSelector(selectors[key] || key);
```

?

(For reference, the specific line I'm referring to is line 2873 of `objects.js` as of `fd2d8df8`)

Making that single-line change "just works" for me, to open things up to use expressions like: `xPosition() / 2 + 10`: 
![image](https://cloud.githubusercontent.com/assets/4712979/21732404/31069540-d40e-11e6-8a18-83158fbb4905.png)

with the caveat that it currently cheerfully allows you to use command blocks in your expressions as well (`turnLeft(15) + 10`, for instance:) 
![image](https://cloud.githubusercontent.com/assets/4712979/21732420/409a2abc-d40e-11e6-9b00-3f4c40643d42.png)

But that could be remedied with a quick additional check for appropriate block types...

I guess one potential issue with this is that not all reporter blocks are actually available via `blockForSelector`; for instance, AFAICT there's no way to get the "tempo" reporter in this way, or (say) "item N of list L". But since we're well into "power user" territory with this feature anyway (though I'm very excited by how much time this can save for power users!), maybe that can just be a caveat. Or maybe it's an excuse to create more tags to get at the core selectors?

Which still leaves the fact that custom reporters won't be selected in this fashion. Which is probably fine, for the same "power user" reasoning?

Anyway, obviously this is a "nice to have", not a "this should really be a part of this", and especially if I'm missing something that makes this much more complex than a simple one-line change + block-type check, this could either become a separate feature request after the current code has been pulled, or just an interesting idea that didn't get implemented. :)

cycomachead commented:
Hmm. I have mixed feelings on going down that route too far. On the one hand, being able to pop simple reporters in place is immensely handy, and I do feel like things should work properly if they are deployed. However, this starts to sound just like a text form of Snap! Which I'm pretty sure we all do not want kids to be using. (And believe me they will try...). 

I wonder if it makes sense to re-think some aspects of using the current text input mechanism to allow building an expression up left to right so that you could do something like insert a block in a slot and then put a new block "under" it so you could build an expression left to right instead of middle out. (I hope that made sense. That was thinking out loud, not recommending that we do such things necessarily.)

-- 
Michael Ball
From my iPhone
michaelball.co

> On Jan 6, 2017, at 12:55 PM, Micah Cowan <notifications@github.com> wrote:
> 
> Right now, function-call identifiers get looked up in a list of reportMonadic variants, and then in a mapping of operator names, which get passed to SpriteMorph.prototype.blockForSelector().
> 
> Are there any thoughts on, if those two checks fail, trying the key/identifier directly in blockForSelector? i.e., in this block of code from blockFromAST():
> 
>         } else { // The function or operator has its own block
>             block = SpriteMorph.prototype.blockForSelector(selectors[key]);
>             inps = block.inputs();
>         }
> the block = line becomes something like:
> 
>             block = SpriteMorph.prototype.blockForSelector(selectors[key] || key);
> ?
> 
> (For reference, the specific line I'm referring to is line 2873 of objects.js as of fd2d8df8)
> 
> Making that single-line change "just works" for me, to open things up to use expressions like: xPosition() / 2 + 10:
> 
> 
> with the caveat that it currently cheerfully allows you to use command blocks in your expressions as well (turnLeft(15) + 10, for instance:)
> 
> 
> But that could be remedied with a quick additional check for appropriate block types...
> 
> I guess one potential issue with this is that not all reporter blocks are actually available via blockForSelector; for instance, AFAICT there's no way to get the "tempo" reporter in this way, or (say) "item N of list L". But since we're well into "power user" territory with this feature anyway (though I'm very excited by how much time this can save for power users!), maybe that can just be a caveat. Or maybe it's an excuse to create more tags to get at the core selectors?
> 
> Which still leaves the fact that custom reporters won't be selected in this fashion. Which is probably fine, for the same "power user" reasoning?
> 
> Anyway, obviously this is a "nice to have", not a "this should really be a part of this", and especially if I'm missing something that makes this much more complex than a simple one-line change + block-type check, this could either become a separate feature request after the current code has been pulled, or just an interesting idea that didn't get implemented. :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

brianharvey commented:
We _do_ want to support "a text form of Snap<i>!</i>," namely the keyboard editor.  We should consider how the math translator fits into that larger project some time after the immediate coming release.

BookOwl commented:
@micahcowan, that's really neat! I think that I'll add it in, if there are no objections from anybody else.

BookOwl committed:
Merge remote-tracking branch 'author/master'
